lib/student_greeter.rb

```diff
@@ -0,0 +1,84 @@
+class StudentGreeter
+  def self.greet_all_ungreeted_students
+    new.greet_all_ungreeted_students
+  end
+
+  def initialize
+    @greeters = User.where(greeter: true)
+  end
+
+  def greet_all_ungreeted_students
+    courses = Course.strictly_current
+    courses.each do |course|
+      nonstudents_in_course = course.nonstudents
+      possible_greeters = @greeters & nonstudents_in_course
+      next if possible_greeters.empty?
+      greeter = possible_greeters[0]
+      course.students_without_nonstudents.each do |student|
```

Comment:
greet_if_ungreeted is still necessary, because a student may have been greeted (manually) even though the flag is not set. So greet_if_ungreeted does a couple of things:
* check for the greeted flag
* check for a blank talk page (which clearly indicates that the user has not been greeted)
* checks if any greeter has posted to the talk page before (and therefore, the user is actually greeted)

The check for the greeted flag can go here, but that will just remove the first line of greet_if_ungreeted.